### PR TITLE
ConvAsm3x3U big tensor support

### LINKED ISSUE #609: ConvAsm3x3U: precision problems
Use case (immediate mode):
```
atamazov@morpheus:~/github/MLOpen2/build/release.hip-clang$ ./bin/MIOpenDriver conv -W 256 -H 256 -c 256 -k 6 -x 3 -y 3 -n 128 -p 1 -q 1 -w 1 -t 1 -i 1 -V 1 $MIO -F 1 -S 0
MIOpenDriver conv -W 256 -H 256 -c 256 -k 6 -x 3 -y 3 -n 128 -p 1 -q 1 -w 1 -t 1 -i 1 -V 1 -C /spb-hlc-fs2/1/users/atamazov/miopen-verify2 -F 1 -S 0
MIOpen(HIP): Info [Handle] stream: 0x2edcdc0, device_id: 0
MIOpen(HIP): Info [GetForwardSolutionCount]
MIOpen(HIP): Info [Measure] Db::Prefetch time: 74.0174 ms
MIOpen(HIP): Info [GetForwardSolutions]
MIOpen(HIP): Info [HipCompilerVersionImpl] 3.5.20214
MIOpen(HIP): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, MIOpen version 2.10.0.8277-2a182a8c6-dirty
Forward Conv solutions available: 3
- id: 1 algo: 1, time: 19.2375 ms, ws: 0, name: ConvAsm3x3U
- id: 14 algo: 3, time: 54.4723 ms, ws: 0, name: ConvBinWinograd3x3U
- id: 33 algo: 0, time: 370.444 ms, ws: 603979776, name: gemm
MIOpen(HIP): Info [GetForwardSolutionWorkspaceSize] solver_id = ConvAsm3x3U
MIOpen(HIP): Info [CompileForwardSolution] solver_id = ConvAsm3x3U
MIOpen(HIP): Info [FindSolutionImpl] ConvAsm3x3U
MIOpen(HIP): Info [SetValues] , content inserted: ConvAsm3x3U:9,6,3
MIOpen(HIP): Info [GetValues] =ConvAsm3x3U:9,6,3
MIOpen(HIP): Info [ConvolutionForwardImmediate] solver_id = ConvAsm3x3U, workspace = 0
Wall-clock Time Forward Conv. Elapsed: 30.7274 ms, Auxiliary API calls: 1652.89 ms (GWSS: 0.0390625)
MIOpen Forward Conv. Algorithm: 1, Solution: 1/ConvAsm3x3U
GPU Kernel Time Forward Conv. Elapsed: 19.460573 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u1, 128, 256, 256, 256, 3, 3, 6,  231928233984, 18446744065119672320, 201326592, 11918, 947903405396, 19.460573
Read data from input file /spb-hlc-fs2/1/users/atamazov/miopen-verify2/conv_fwd_out_0_2_0_1_128x256x256x256_6x256x3x3_1x1_1x1_1x1_0x0_0_0_GPUfloat_REFfloat
Forward Convolution Failed: 0.0994822 > 1.5e-06
```

### PR DESCRIPTION
**Pull request by Slimakanzer**
This PR expands the range of valid values for i/o tensors of ConvAsm3x3U solver. Also resolves _issue_ #609 ~issue~ and removes temporary limits introduced in #643 PR.

## Summary of code changes
* Updated *ConvAsm3x3U* kernel with big i/o tensors support
* Updated *ConvAsm3x3U* limitations in `IsApplicable()`

## Local verification
* gfx906, ROCm 3.5.1
* OCL + HIP backend

W | H | c | n | k | x | y | p | q | u | v | g | dir
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- 
128 | 512 | 128 | 160 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
128 | 512 | 128 | 160 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
128 | 512 | 128 | 160 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | fwd
224 | 224 | 32 | 128 | 224 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
224 | 224 | 32 | 128 | 224 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
224 | 224 | 32 | 128 | 224 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | fwd
224 | 224 | 32 | 128 | 224 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | bwd
512 | 256 | 96 | 128 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
512 | 256 | 96 | 128 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
512 | 256 | 96 | 128 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | fwd
256 | 256 | 256 | 96 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
256 | 256 | 256 | 96 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
256 | 256 | 256 | 96 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | fwd
256 | 256 | 256 | 96 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | bwd
32| 32 | 256 | 4100 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
32| 32 | 256 | 4100 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
32| 32 | 256 | 4100 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | fwd
32| 32 | 256 | 4100 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | bwd
64| 64 | 256 | 1040 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
64| 64 | 256 | 1040 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
64| 64 | 256 | 1040 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | fwd
64| 64 | 256 | 1040 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | bwd
24| 24 | 512 | 3712 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
24| 24 | 512 | 3712 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
24| 24 | 512 | 3712 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | fwd
24| 24 | 512 | 3712 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | bwd
96| 96 | 128 | 960 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
96| 96 | 128 | 960 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
96| 96 | 128 | 960 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | fwd
96| 96 | 128 | 960 | 16 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | bwd
14| 14 | 512 | 11264 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
14| 14 | 512 | 11264 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
14| 14 | 512 | 11264 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | fwd
14| 14 | 512 | 11264 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | bwd
56| 56 | 640 | 576 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
56| 56 | 640 | 576 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
56| 56 | 640 | 576 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | fwd
56| 56 | 640 | 576 | 32 | 3 | 3 | 1 | 1 | 1 | 1 | 4 | bwd
56| 56 | 256 | 1408 | 64 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd
56| 56 | 256 | 1408 | 64 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd
56| 56 | 256 | 1408 | 64 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | fwd
56| 56 | 256 | 1408 | 64 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | bwd

## Local performance testing
No performance regression found. This table shows performance improvements compared with current develop branch.
* gfx906, ROCm 3.5.1
* OCL backend

W | H | c | n | k | x | y | p | q | u | v | g | dir | | perf
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
64 | 128 | 64 | 64 | 256 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd | | 1%
64 | 128 | 64 | 64 | 256 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd | | 0%
64 | 128 | 64 | 64 | 256 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | fwd | | -1%
64 | 128 | 64 | 64 | 256 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | bwd | | 0%
64 | 512 | 64 | 128 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd | | 0%
64 | 512 | 64 | 128 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd | | 3%
64 | 512 | 64 | 128 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | fwd | | 2%
128 | 128 | 64 | 64 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd | | 3%
128 | 128 | 64 | 64 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd | | -2%
128 | 128 | 64 | 64 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | fwd | | 0%
256 | 128 | 64 | 64 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | fwd | | 0%
256 | 128 | 64 | 64 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 1 | bwd | | 0%
256 | 128 | 64 | 64 | 8 | 3 | 3 | 1 | 1 | 1 | 1 | 8 | fwd | | 3%

### CODE REVIEW

File: src/solver/conv_asm_3x3u.cpp [hunk 1]
```diff
@@ -209,8 +209,11 @@ bool ConvAsm3x3U::IsApplicable(const ConvolutionContext& params) const
         && (params.n_inputs / params.group_counts) % 4 == 0 /// \todo: remove restriction that (n_inputs/group_counts) must be multiple of 4
         && params.in_width > 3
         && params.in_width <= 1000
-        && IN_BUF_SZ <= 4 * GIB
-        && OUT_BUF_SZ <= 4 * GIB
+        && h_w < std::pow(2, 29)
```
Comment (Kirpich30000) [inline src/solver/conv_asm_3x3u.cpp:212]: `filters_size` should be less than 4GB

**Comment by atamazov** [comment]
@Slimakanzer Have you tested correctness with 9 configs only?

Have you tested how this PR affects performance?

File: src/kernels/conv3x3.s [hunk 2]
```diff
@@ -58,29 +58,10 @@ gid_z = 4
 // dwords 0:1 - input buffer pointer
 // dwords 2:3 - weights pointer
 // dwords 4:5 - output buffer pointer
-// dwords 6:7 - debug buffer pointer
 .set in_ptr_off, 0x0
 .set wei_ptr_off, 0x8
 .set out_ptr_off, 0x10
 
-.ifnotdef no_params_file
```
Comment (atamazov) [inline src/kernels/conv3x3.s:66]: `no_params_file` must be removed from the solver.

File: src/kernels/conv3x3.s [hunk 3]
```diff
@@ -220,8 +201,7 @@ output_buffer_window = 4 * img_height * img_width
 .if limit_wave_cnt
   .SET_MAX_WAVES_LIMIT limit_wave_cnt
 .endif
-//xnack disabled by default
-//.SGPR_RESERVE_XNACK
+.SGPR_RESERVE_XNACK
```
Comment (atamazov) [inline src/kernels/conv3x3.s:204]: [Q] why this change?

File: src/kernels/conv3x3.s [hunk 4]
```diff
@@ -729,32 +681,49 @@ miopenGcnAsmConv3x3U:
       s_addc_u32 s[weights_ptr+1], s[weights_ptr+1], 0
   .endif
 
+.macro s_mul_u64_u32 dst_lo, dst_hi, src0, src1, vtmp0, vtmp1
```
Comment (atamazov) [inline src/kernels/conv3x3.s:684]: [Recommendation]
```suggestion
.macro s_mul_u64_u32 sdst_lo, sdst_hi, ssrc0, ssrc1, vtmp0, vtmp1
```

File: src/kernels/conv3x3.s [hunk 5]
```diff
@@ -729,32 +681,49 @@ miopenGcnAsmConv3x3U:
...
 .if group_counts > 1
-  s_mul_i32 s[tmp], s[s_group_id], c_group_size * input_feature_map_stride // c_group_offset
-  s_add_u32 s[in_desc], s[in_desc], s[tmp]
-  s_addc_u32 s[in_desc+1], s[in_desc+1], 0
+    // adds (s[s_group_id] * input_feature_map_stride * c_group_size)
+    // to input base addr
+    s_mul_i32  s[stmp2], s[s_group_id], c_group_size
+    s_add_u32 s[tmp], s[tmp], s[stmp2]
+.endif
+.if batch_size > 1 || group_counts > 1
```
Comment (atamazov) [inline src/kernels/conv3x3.s:706]: Configs with `group_count != 1` must be tested.

File: src/solver/conv_asm_3x3u.cpp [hunk 6]
```diff
@@ -190,12 +190,14 @@ bool ConvAsm3x3U::IsApplicable(const ConvolutionContext& params) const
         return false;
     }
 
-    constexpr auto GIB     = static_cast<int64_t>(1024) * 1024 * 1024;
```
Comment (atamazov) [inline src/solver/conv_asm_3x3u.cpp:193]: What is wrong with GIB?

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/solver/conv_asm_3x3u.cpp (hunk 1 above)
Comment (atamazov) [reply src/solver/conv_asm_3x3u.cpp:214]: I guess you can find this check at line 219

File: src/kernels/conv3x3.s (hunk 5 above)
Comment (Slimakanzer) [reply src/kernels/conv3x3.s:706]: Thank you! You can find these test cases in the updated PR description.

File: src/kernels/conv3x3.s (hunk 4 above)
Comment (Slimakanzer) [reply src/kernels/conv3x3.s]: Thanks for the recommendation, fixed in bd0f6aa

File: src/kernels/conv3x3.s (hunk 2 above)
Comment (Slimakanzer) [reply src/kernels/conv3x3.s:66]: Thanks for the quality review, fixed in c41f359

File: src/solver/conv_asm_3x3u.cpp (hunk 6 above)
Comment (Slimakanzer) [reply src/solver/conv_asm_3x3u.cpp:193]: Everything is okay, i tried to make it look like in other solvers, fixed in 7ddf090

File: src/solver/conv_asm_3x3u.cpp [hunk 7]
```diff
@@ -209,8 +213,12 @@ bool ConvAsm3x3U::IsApplicable(const ConvolutionContext& params) const
         && (params.n_inputs / params.group_counts) % 4 == 0 /// \todo: remove restriction that (n_inputs/group_counts) must be multiple of 4
         && params.in_width > 3
         && params.in_width <= 1000
-        && IN_BUF_SZ <= 4 * GIB
-        && OUT_BUF_SZ <= 4 * GIB
+        && ELEM_SZ * h_w < GIB
+        && ELEM_SZ * n_c < 4 * GIB
+        && ELEM_SZ * n_k < 4 * GIB
+        && ELEM_SZ * n_c_h_w < 256 * TIB
+        && ELEM_SZ * n_k_h_w < 256 * TIB
```
Comment (atamazov) [inline src/solver/conv_asm_3x3u.cpp:220]: Original `OUT_BUF_SZ` does not assume that `(out_H == in_H) && (out_W == in_W)` and looks clearer, right?

**Comment by atamazov** [comment]
@Slimakanzer Are you pretty sure that 15 test cases is enough for correctness testing? Or actual amount of testing is somewhat bigger?

File: src/solver/conv_asm_3x3u.cpp (hunk 7 above)
Comment (Slimakanzer) [reply src/solver/conv_asm_3x3u.cpp]: You are right, thanks for the recommendation!

**Comment by Slimakanzer** [comment]
@atamazov Updated list of verified cases. The list contains only big-tensors. A lot of small tensors have also been tested. Cases were verified only on **_gfx9_** and verification was not performed at **_gfx8_**, however, only instructions were used that are in _gfx9 isa_ and _gfx8 isa_.

File: src/solver/conv_asm_3x3u.cpp [hunk 8]
```diff
@@ -209,8 +214,13 @@ bool ConvAsm3x3U::IsApplicable(const ConvolutionContext& params) const
         && (params.n_inputs / params.group_counts) % 4 == 0 /// \todo: remove restriction that (n_inputs/group_counts) must be multiple of 4
         && params.in_width > 3
         && params.in_width <= 1000
-        && IN_BUF_SZ <= 4 * GIB
-        && OUT_BUF_SZ <= 4 * GIB
+        && IN_IMG_SZ < GIB
```
Comment (atamazov) [inline src/solver/conv_asm_3x3u.cpp:217]: I think that `<=` should be used everywhere.

**Review by atamazov — CHANGES_REQUESTED** [review]
Almost done!

File: src/solver/conv_asm_3x3u.cpp [hunk 9]
```diff
@@ -209,8 +214,13 @@ bool ConvAsm3x3U::IsApplicable(const ConvolutionContext& params) const
...
+        && OUT_IMG_SZ < 4 * GIB
+        && IN_FEATURE_COUNT  < 4 * GIB
```
[the remark this answers is not in the document]
Comment (Slimakanzer) [reply src/solver/conv_asm_3x3u.cpp]: Looked carefully at the shader source code and you're right, thanks!

**Review by atamazov — APPROVED** [review]
LGTM!

File: src/kernels/conv3x3.s (hunk 3 above)
Comment (atamazov) [reply src/kernels/conv3x3.s:204]: I asked and forgot. You have not answered. But this is the cause of the problem in amd-develop that led to the failed testing round in the staging. Let's be more attentive next time.

I will prepare hotfix.

Comment (atamazov) [reply src/kernels/conv3x3.s:204]: @Slimakanzer Please answer my initial question, thanks.
